### PR TITLE
Speedometer 3: WebContent spends a lot of time in RemoteImageBufferProxy::ensureBackendCreated.

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### cab67b0bc9ac57b0ee056b50359cff8d5e0f260f
<pre>
Speedometer 3: WebContent spends a lot of time in RemoteImageBufferProxy::ensureBackendCreated.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266728">https://bugs.webkit.org/show_bug.cgi?id=266728</a>
&lt;<a href="https://rdar.apple.com/117823468">rdar://117823468</a>&gt;

Reviewed by NOBODY (OOPS!).

Synchronously waiting on the ImageBufferBackendHandle for an ImageBufferSet can block the main thread while
trying to serialize a transaction.

This introduces a new message receiver &apos;RemoteRenderingBackendProxyBackground&apos;, which RemoteRenderingBackendProxy
listens for on a background work queue. It moves the reply to &apos;prepareImageBufferSetsForDisplay&apos; to be a message
on the background receiver.

RemoteLayerTree transaction serialization on the main thread now just serializes an identifier for the image
buffer set and doesn&apos;t block.

The transaction flusher thread waits for any incoming responses on the remote rendering background thread, and
builds a hashamp of image buffer set identifier to response data and appends this on to the remote layer tree
transaction.

RemoteLayerTreeDrawingAreaProxy uses the trailing hashmap to copy across any needed backend handles from the
response hashmap into the appropriate RemoteLayerBackingStoreProperties.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplaySync):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
(WebKit::RemoteRenderingBackend::identifier):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(class_template_headers):
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.h:
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::bufferSetIdentifier const):
(WebKit::RemoteLayerBackingStoreProperties::bufferSetIdentifier):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStoreProperties::decode):
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::setBackendHandle):
(WebKit::RemoteLayerBackingStore::takePendingFlushers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerBackingStoreCollection::didFlushLayers):
(WebKit::RemoteLayerBackingStoreCollection::backingStoreWillBeEncoded): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlusher):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlushers): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createFlusher):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::bufferSetIdentifier const):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createFlushers): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxy::flushFrontBufferAsync):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
(WebKit::RemoteRenderingBackendProxy::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackendProxyBackground::imageBufferSetDidPrepareForDisplay):
(WebKit::RemoteRenderingBackendProxyBackground::takeImageBufferSetPrepareForDisplayOutputData):
(WebKit::RemoteRenderingBackendProxy::ensurePrepareCompleted): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::background const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxyBackground.h: Added.
(WebKit::RemoteRenderingBackendProxyBackground::ImageBufferSetPrepareBufferForDisplayOutputDataWrapper::ImageBufferSetPrepareBufferForDisplayOutputDataWrapper):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxyBackground.messages.in: Copied from Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::BackingStoreFlusher::flush):
</pre>
https://github.com/WebKit/WebKit/commit/cab67b0bc9ac57b0ee056b50359cff8d5e0f260f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34326 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13132 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36317 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37004 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31064 "Hash cab67b0b for PR 22006 does not build (failure)") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10245 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30073 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34838 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11127 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30608 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9696 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/34623 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9807 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38288 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31192 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30975 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35869 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9919 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7793 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33800 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11715 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10476 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10756 "Built successfully") | | |

### CODE REVIEW

**Comment by fujii** [comment]
Thank you very much for taking care of Windows port. You saved my holidays.

File: Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp [hunk 1]
```diff
@@ -100,6 +104,8 @@ void RemoteRenderingBackendProxy::ensureGPUProcessConnection()
         // connection. This prevents waits on RemoteRenderingBackendProxy to process messages from other connections.
         m_streamConnection->open(*this, *this);
 
+        m_streamConnection->connectionForTesting().addWorkQueueMessageReceiver(Messages::RemoteRenderingBackendProxyBackground::messageReceiverName(), m_queue, m_background.get(), m_parameters.identifier.toUInt64());
```
Comment (kkinnunen-apple) [inline Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:107]: this should be `m_streamConnection->addWorkQueueMessageReceiver`
E.g. eventually StreamClientConnection should implement the IPC::Connection interface (if needed).

File: Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxyBackground.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    class ImageBufferSetPrepareBufferForDisplayOutputDataWrapper : public ThreadSafeRefCounted< ImageBufferSetPrepareBufferForDisplayOutputDataWrapper > {
+    public:
+        ImageBufferSetPrepareBufferForDisplayOutputDataWrapper(ImageBufferSetPrepareBufferForDisplayOutputData&& data)
+            : m_outputData(WTFMove(data))
+        { }
+
+        ImageBufferSetPrepareBufferForDisplayOutputData m_outputData;
+    };
+
+    IPC::ThreadSafeObjectHeap<RemoteImageBufferSetIdentifier, RefPtr< ImageBufferSetPrepareBufferForDisplayOutputDataWrapper>> m_imageBufferSetsOutputData;
```
Comment (kkinnunen-apple) [inline Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxyBackground.h:57]: This looks odd, as there's no-one giving out versioned ids. Not sure if this is needed..

Comment (mattwoodrow) [reply Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxyBackground.h:57]: What we need (I believe), is a hash table, a lock, and a condition variable, along with code that can use these to wait until a given key is added to the hash map.

`IPC::ThreadSafeObjectHeap` provides that (along with more), but I don't think we currently have any other helpers for this. I can of course just reimplement.

**Pull request closed without merging**